### PR TITLE
[SYCL][V++] Add kernel v++ arg forwarding

### PR DESCRIPTION
**Pull request by Ralender**
prototype implementation of forwarding v++ arguments for a kernel.

### CODE REVIEW

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+int main() {
+  sycl::buffer<sycl::cl_int, 1> Buffer(4);
+  sycl::queue Queue;
+  sycl::range<1> NumOfWorkItems{Buffer.get_count()};
+
+  Queue.submit([&](sycl::handler &cgh) {
+    auto Accessor = Buffer.get_access<sycl::access_mode::write>(cgh);
+    cgh.single_task<class FirstKernel>([=]() {
+      "--optimize 2"_kernel_param;
```
Comment (keryell) [inline sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: Warf! :-)

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 2]
```diff
@@ -25,6 +25,33 @@
...
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+namespace  {
+template<char...Chars>
+constexpr const char FixedString[] = {Chars...};
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-string-literal-operator-template"
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: What is the problem here?
At least mention it in a comment.
I think this should be compiled in C++20 mode anyway.

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      // CHECK: v++ {{.*}}class_FirstKernel{{.*}} --optimize 2
+      Accessor[0] = 0;
+    });
+  });
+
+  Queue.submit([&](sycl::handler &cgh) {
+    auto Accessor = Buffer.get_access<sycl::access_mode::write>(cgh);
+    cgh.single_task<class SecondKernel>([=]() {
```
Comment (keryell) [inline sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: As an alternative syntax, it would be nice to have also a decorator syntax to the kernel, etc.

File: sycl/tools/sycl-xocc/bin/sycl-xocc [hunk 4]
```diff
@@ -211,23 +211,23 @@ if [[ ! ${#KERNEL_NAME_ARRAY[@]} -eq 0 ]]; then
   # Compile all Kernels in the source file to seperate .xo ("object") files
   # and add them to the list of kernels to be linked
   LINKER_LIST=""
-  for index in "${KERNEL_NAME_ARRAY[@]}"; do
-  if [[ -n "$index" ]]; then
+  for ((i=0;i<${#KERNEL_NAME_ARRAY[@]};++i)); do
```
Comment (keryell) [inline sycl/tools/sycl-xocc/bin/sycl-xocc:214]: What was the issue here?

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp (hunk 1 above)
Comment (keryell) [reply sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: It looks like Synopsys CHESS random words... :-)

File: llvm/lib/SYCL/KernelPropGen.cpp [hunk 5]
```diff
@@ -124,13 +125,71 @@ struct KernelPropGen : public ModulePass {
     return 0;
   }
 
+  void AddExtraArgsToCallers(Function *Original, std::string Additional) {
```
Comment (keryell) [inline llvm/lib/SYCL/KernelPropGen.cpp:130]: All this require some comments and even high-level design poetry.

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp (hunk 2 above)
Comment (keryell) [reply sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: And by the way, as usual, add a lot of comments all over the place.

File: sycl/tools/sycl-xocc/bin/sycl-xocc [hunk 6]
```diff
@@ -211,23 +211,23 @@ if [[ ! ${#KERNEL_NAME_ARRAY[@]} -eq 0 ]]; then
...
     fi
 
     $SDX --target $XCL_EMULATION_MODE --platform $XILINX_PLATFORM \
       --xp param:compiler.hlsDataflowStrictMode=off \
-      --save-temps -c -k "$index" -o "${TMP_PATH}_$index.xo" \
+      --save-temps -c -k "${KERNEL_NAME_ARRAY[i]}" -o "${TMP_PATH}_${KERNEL_NAME_ARRAY[i]}.xo" \
       "${TMP_PATH}_${SOURCE_FILE_NAME}_kernels-linked.opt.xpirbc" \
-      $ADDITIONAL_COMPILE_ARGS &
-      LINKER_LIST="${TMP_PATH}_$index.xo $LINKER_LIST"
+      $ADDITIONAL_COMPILE_ARGS ${EXTRA_KERNEL_PARAMS_ARRAY[i]} &
```
Comment (keryell) [inline sycl/tools/sycl-xocc/bin/sycl-xocc:229]: I completely missed this variable at first reading...
A lot of comments...

File: llvm/lib/SYCL/KernelPropGen.cpp [hunk 7]
```diff
@@ -184,6 +243,11 @@ struct KernelPropGen : public ModulePass {
               "# xocc linker phase\n";
        O << "DDR_BANK_ARGS=\"" << DDRArgs.str() << "\"\n";
     }
+
+    if (!ExtraArgs.empty()) {
```
Comment (keryell) [inline llvm/lib/SYCL/KernelPropGen.cpp:254]: Comments...
Explain the Bash ABI here. :)

**Review by keryell — CHANGES_REQUESTED** [review]
This allows quite a lot of fun.
I cannot wait for `_vhdl` and `_verilog`. :-)
Just a few COMMENTS...

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp (hunk 2 above)
Comment (Ralender) [reply sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: this is not using C++20 but a compiler extension from clang and gcc, because this is only needed for device code using clang compiler extensions is fine. but this is to silence the warning about that compiler extension.

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp (hunk 1 above)
Comment (Ralender) [reply sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: i absolutely agree that the API is not yet so good, but just getting a way to provide constant string to the annotation was already hard because clang doesn't yet fully support arbitrary non-type template parameters of C++20. so i am using a compiler extension. the compiler extension i am using is quite powerful and i have some ideas to improve the API

File: sycl/tools/sycl-xocc/bin/sycl-xocc (hunk 4 above)
Comment (Ralender) [reply sycl/tools/sycl-xocc/bin/sycl-xocc:214]: the is no issue in the previous implemenantation. but now we need to iterate on the KERNEL_NAME_ARRAY and EXTRA_KERNEL_PARAMS_ARRAY so we can't use the old system.

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp (hunk 2 above)
Comment (Ralender) [reply sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp:39]: This is a compiler extension not C++20 so this doesn't need C++20

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 8]
```diff
@@ -21,35 +21,166 @@
...
+template <auto N, typename Base> struct StrNumber<true, number<N, Base>> {
+  using last_didgit = typename Base::template at<N % Base::size>;
+  using type = typename concat<
+      cstr<char>,
+      typename StrNumber<(N / Base::size >= Base::size), number<N / Base::size, Base>>::type,
+      last_didgit>::type;
+};
+
+template <auto N, typename Base> struct StrNumber<false, number<N, Base>> {
+  using type = typename Base::template at<N % 10>;
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: Why not `N % Base::size`?

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 9]
```diff
@@ -21,35 +21,166 @@
...
+        return std::forward<Func>(internal_f)(
+            std::forward<decltype(args)>(args)...);
+      });
+}
+
+template <typename... Params, size_t... ParamIdx>
+auto kernel_param(std::tuple<Params...> params,
+                  std::index_sequence<ParamIdx...>) {
+  return detail::kernel_param_impl(std::get<sizeof...(Params) - 1>(params),
+                            std::get<ParamIdx>(params)...);
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: Indent

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 10]
```diff
@@ -21,35 +21,166 @@
...
+      typename StrNumber<(N / Base::size >= Base::size), number<N / Base::size, Base>>::type,
+      last_didgit>::type;
+};
+
+template <auto N, typename Base> struct StrNumber<false, number<N, Base>> {
+  using type = typename Base::template at<N % 10>;
+};
+#endif
+
+template <typename Func, typename... CharPacks>
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: Comment...

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 11]
```diff
@@ -21,35 +21,166 @@
...
+auto kernel_param_impl(Func &&f, CharPacks...) {
+  return sycl::detail::make_const_cheater(
+      [internal_f = std::forward<Func>(f)](auto &&... args) mutable {
+        __SYCL_DEVICE_ANNOTATE(
+            "xilinx_kernel_param",
+            sycl::detail::StrPacker<typename sycl::detail::concat<
+                sycl::detail::cstr<char, ' '>, CharPacks...>::type>::Str)
+        int useless;
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: Comment...

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp [hunk 12]
```diff
@@ -19,19 +19,20 @@ int main() {
 
   Queue.submit([&](sycl::handler &cgh) {
     auto Accessor = Buffer.get_access<sycl::access_mode::write>(cgh);
-    cgh.single_task<class FirstKernel>([=]() {
-      "--optimize 2"_kernel_param;
-      // CHECK: v++ {{.*}}class_FirstKernel{{.*}} --optimize 2
-      Accessor[0] = 0;
-    });
+    cgh.single_task<class FirstKernel>(
+        sycl::xilinx::kernel_param("--optimize 2"_cstr, [=]() {
```
Comment (keryell) [inline sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]:
```suggestion
        sycl::xilinx::kernel_param("--optimize 2"_cstr, [=] {
```

Comment (keryell) [reply sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: At the end I wonder why not using just
```c++
sycl::xilinx::kernel_param<"--optimize 2">([=] {
```
that work in C++20 I think.

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp [hunk 13]
```diff
@@ -19,19 +19,20 @@ int main() {
...
   Queue.submit([&](sycl::handler &cgh) {
     auto Accessor = Buffer.get_access<sycl::access_mode::write>(cgh);
-    cgh.single_task<class SecondKernel>([=]() {
-      "--kernel_frequency 300"_kernel_param;
-      // CHECK: v++ {{.*}}class_SecondKernel{{.*}} --kernel_frequency 300
-      Accessor[1] = 1;
-    });
+    cgh.single_task<class SecondKernel>(sycl::xilinx::kernel_param(
+        "--kernel_frequency"_cstr,
+        sycl::xilinx::number<0x100 + 0x200, sycl::detail::Base16>::str, [=]() {
```
Comment (keryell) [inline sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]:
```suggestion
        sycl::xilinx::number<0x100 + 0x200, sycl::detail::Base16>::str, [=] {
```

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp (hunk 2 above)
Comment (keryell) [reply sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: All this requires even more comments and to be moved into a separate file higher in the hierarchy, the equivalent of `include/triSYCL/detail/...`.

**Review by keryell — COMMENTED** [review]
Is this really reasonable? :-)

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+///  \file
+///  This file contains meta programming utilities to manipulate string
+///  constants Stored in a type like cstr<char, 'a', 'b'>. This is used for
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
///  constants stored in a type like cstr<char, 'a', 'b'>. 
///  This can be used for example with
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+///  kernel properties because they must be constants and the API reflects that.
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
///  kernel properties because they must be constant and the API reflects that.
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+/// utility to transform a string literal into its cstr form.
+template <typename CharT, CharT... charpack>
+__SYCL_ALWAYS_INLINE constexpr cstr<CharT, charpack...>
+operator"" _cstr() noexcept {
+  return {};
+}
+#pragma clang diagnostic pop
+
+using Base2 = decltype("01"_cstr);
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: Perhaps you can coin some design comments here. Like this has 3 roles:
- as a tag to select a base;
- as a description of each character to use for each "digit";
- as the base itself by just counting the number of digit characters.

Do not forget the readers are not telepathic or you. :-)

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+using Base16 = decltype("0123456789abcdef"_cstr);
+
+#else
+__SYCL_ALWAYS_INLINE constexpr cstr<char> operator"" _cstr(const char *,
+                                                           size_t) noexcept {
+  return {};
+}
+
+/// When not in device mode the kernel properties are not used so they do not
+/// need to be correct just to compile correctly.
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: So making all a base 1 is fine :-)

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+#include "CL/sycl/detail/pi.h"
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+
+namespace sycl {
+
+namespace detail {
+
+/// utility to make any functor's operator() be const even if it originally
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: I have the feeling this is yet another utility that should go in another file.

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+template <typename CharT, CharT... Sep, CharT... P1, CharT... P2,
+          typename... Rest>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>, cstr<CharT, P2...>,
+              Rest...> {
+  using type =
+      typename concat<cstr<CharT, Sep...>, cstr<CharT, P1..., Sep..., P2...>,
+                      Rest...>::type;
+};
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: I have the feeling you start reimplementing Boost.Hana here...
But no obvious solution...

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+/// transform a type like: cstr<char, 'a', 'b'>
+/// into a string "ab"
+template <typename> struct StrPacker {};
+
+/// utility type to manipulate string in constant context
+template <typename CharT, CharT... charpack> struct cstr {
+  using Char = CharT;
+  static constexpr unsigned size = sizeof...(charpack);
+  template <unsigned Idx>
+  using at = __type_pack_element<Idx, cstr<CharT, charpack>...>;
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: You cannot write such a line uncommented! :-(
You could explain what it does and add some links to the GCC & Clang extensions.

File: sycl/include/CL/sycl/detail/cstr.hpp (hunk 19 above)
Comment (keryell) [reply sycl/include/CL/sycl/detail/cstr.hpp]: And probably what we want is also a `concat_t` alias or perhaps even a `concat_v` `constexpr` function...

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp [hunk 21]
```diff
@@ -29,121 +30,15 @@ namespace sycl{
...
+/// kernel properties that can be pickup by the KernelPropGen pass.
 template <typename Func, typename... CharPacks>
 auto kernel_param_impl(Func &&f, CharPacks...) {
   return sycl::detail::make_const_cheater(
       [internal_f = std::forward<Func>(f)](auto &&... args) mutable {
+        /// Generate an annotation with as argument all the provided properties
+        /// concatenated and separeted by a space.
         __SYCL_DEVICE_ANNOTATE(
             "xilinx_kernel_param",
             sycl::detail::StrPacker<typename sycl::detail::concat<
```
Comment (keryell) [inline sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]:
```suggestion
            sycl::detail::StrPacker<typename sycl::detail::concat_t<
```
and remove `::type`

**Review by keryell — CHANGES_REQUESTED** [review]
Besides these remarks, there are still a lot of old comment requests I made which are not addressed yet...

**Comment by keryell** [comment]
Still some stuff here...

File: sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp (hunk 12 above)
Comment (Ralender) [reply sycl/test/xocc_tests/simple_tests/kernel_arguement_forwarding.cpp]: This is not yet fully implemented in clang.
but it should be possible when it is.

File: sycl/include/CL/sycl/detail/cstr.hpp (hunk 19 above)
Comment (Ralender) [reply sycl/include/CL/sycl/detail/cstr.hpp]: yeah this is reimplemeting generic meta programming utilities

File: sycl/include/CL/sycl/detail/cstr.hpp (hunk 20 above)
Comment (Ralender) [reply sycl/include/CL/sycl/detail/cstr.hpp]: i would have considered the names to be clear enough

File: llvm/lib/SYCL/KernelPropGen.cpp [hunk 22]
```diff
@@ -124,13 +125,75 @@ struct KernelPropGen : public ModulePass {
     return 0;
   }
 
+  /// Add the provided string as argument to all kernel function that can reach
+  /// the Original function.
```
Comment (keryell) [inline llvm/lib/SYCL/KernelPropGen.cpp]:
```suggestion
  /// Add the provided string as argument to all kernel functions that can reach
  /// the original function.
```

File: llvm/lib/SYCL/KernelPropGen.cpp [hunk 23]
```diff
@@ -124,13 +125,75 @@ struct KernelPropGen : public ModulePass {
...
+          Stack.append(U->user_begin(), U->user_end());
+          continue;
+        }
+        if (auto *CB = dyn_cast<CallBase>(U))
+          if (CB->getIntrinsicID() == Intrinsic::var_annotation) {
+            Constant *Args =
+                (cast<GlobalVariable>(getUnderlyingObject(CB->getOperand(4)))
+                     ->getInitializer());
+            if (auto *C = dyn_cast<ConstantStruct>(Args)) {
+              std::string Args;
```
Comment (keryell) [inline llvm/lib/SYCL/KernelPropGen.cpp]: It becomes hard to follow if you mask variables like this... `ArgsStr`?

File: llvm/lib/SYCL/KernelPropGen.cpp [hunk 24]
```diff
@@ -184,6 +247,13 @@ struct KernelPropGen : public ModulePass {
               "# xocc linker phase\n";
        O << "DDR_BANK_ARGS=\"" << DDRArgs.str() << "\"\n";
     }
+
+    /// output the list of user specified extra parameter that need to be
```
Comment (keryell) [inline llvm/lib/SYCL/KernelPropGen.cpp]:
```suggestion
    /// Output the list of user specified extra parameter that need to be
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+#include "CL/sycl/detail/pi.h"
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+
+namespace sycl {
+
+namespace detail {
+
+/// utility to make any functor's operator() be const even if it originally
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Utility to make any functor's operator() be const even if it originally
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+  }
+  template <typename... Ts> auto operator()(Ts &&... ts) const && {
+    return std::forward<T &&>(t)(std::forward<Ts>(ts)...);
+  }
+};
+template <typename T> auto make_const_cheater(T &&t) {
+  return const_cheating_wrapper<T>{std::forward<T>(t)};
+}
+
+/// transform a type like: cstr<char, 'a', 'b'>
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Transform a type like: cstr<char, 'a', 'b'>
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+/// into a string "ab"
+template <typename> struct StrPacker {};
+
+/// utility type to manipulate string in constant context
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Utility type to manipulate string in constant context
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+template <typename CharT, CharT... charpack> struct cstr {
+  using Char = CharT;
+  static constexpr unsigned size = sizeof...(charpack);
+  /// access character at position Idx in the string
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
  /// Access character at position Idx in the string
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+  template <unsigned Idx>
+  using at = __type_pack_element<Idx, cstr<CharT, charpack>...>;
+};
+
+template <typename CharT, CharT... P1> struct StrPacker<cstr<CharT, P1...>> {
+  static constexpr const CharT Str[] = {P1...};
+};
+
+/// utility to merge string with a separator
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Utility to merge string with a separator
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+using concat_t = typename concat<Ts...>::type;
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+/// We are using a compiler extention here and we don't want any warning about
+/// it so we suppress that warning.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-string-literal-operator-template"
+
+/// utility to transform a string literal into its cstr form.
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Utility to transform a string literal into its cstr form.
```
Start sentences with uppercase!

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+/// When not in device mode the kernel properties are not used so they do not
+/// need to be correct just to compile correctly.
+using Base2 = cstr<char, '0'>;
+using Base10 = cstr<char, '0'>;
+using Base16 = cstr<char, '0'>;
+
+#endif
+
+#ifdef __SYCL_DEVICE_ONLY__
+/// utility to transform a number constant like 348 into
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]:
```suggestion
/// Utility to transform a number constant like 348 into
```

File: sycl/include/CL/sycl/detail/cstr.hpp [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+template <auto N, typename Base> struct StrNumber<true, number<N, Base>> {
+  using last_didgit = typename Base::template at<N % Base::size>;
+  using type =
+      concat_t<cstr<char>,
+                      typename StrNumber<(N / Base::size >= Base::size),
+                                         number<N / Base::size, Base>>::type,
+                      last_didgit>;
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/cstr.hpp]: Curious indent

File: sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp (hunk 11 above)
Comment (keryell) [reply sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp]: So I still do not know why you need this useless variable since there is no comment...

**Review by keryell — CHANGES_REQUESTED** [review]
Getting closer...

File: llvm/lib/SYCL/KernelPropGen.cpp (hunk 23 above)
Comment (Ralender) [reply llvm/lib/SYCL/KernelPropGen.cpp]: agreed, I didn't realized i reused the same name.

**Comment by keryell** [comment]
I removed the `[WIP]` but I have a doubt... Is it still WIP or ready to merge?

**Comment by Ralender** [comment]
> I removed the `[WIP]` but I have a doubt... Is it still WIP or ready to merge?

yeah its not WIP anymore, i think it is ready to merge.
